### PR TITLE
Optimized templated attention to use exp2

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124444
* #124443
* __->__ #124356

0.705 (vs. FA2) to 0.860 after this change.

image

to

image

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: benchmarks/transformer/score_mod.py [hunk 1]
```diff
@@ -198,15 +202,7 @@ def relative_bias(score, b, h, m, n):
 
     def head_bias(score, b, h, m, n):
         return score + 2 * h
-
```
Comment (drisspg) [inline benchmarks/transformer/score_mod.py:201]: Anti pathogical?

File: test/inductor/test_templated_attention.py [hunk 2]
```diff
@@ -49,18 +49,25 @@ def run_test(self, score_mod: Callable, dtype: torch.dtype = torch.float16):
...
-        tolerance = Tolerances(atol=2e-2, rtol=2e-2)
-        torch.testing.assert_close(
-            ref_out.to(dtype=torch.float32),
-            compiled_out.to(dtype=torch.float32),
-            atol=tolerance.atol,
-            rtol=tolerance.rtol,
-        )
+        compiled_error = (golden_out - compiled_out).abs().mean()
+        ref_error = (golden_out - ref_out).abs().mean()
+        if compiled_error > ref_error * 1.1:
```
Comment (drisspg) [inline test/inductor/test_templated_attention.py:60]: yup, good change this how we do the rest of the sdpa tests

Comment (drisspg) [reply test/inductor/test_templated_attention.py:60]: https://github.com/pytorch/pytorch/blob/c9ab9248ce0391c233e7189c0ee71f2aca6a786c/test/test_transformers.py#L93

File: torch/_inductor/kernel/templated_attention.py [hunk 3]
```diff
@@ -49,10 +49,8 @@ def sdpa_grid(batch_size, num_heads, num_queries, d_model, meta):
     H = {{size("Q", 1)}}
     N_CTX = {{size("Q", 2)}}
 
-    # TODO I think we should do some performance work
-    # to find the optimal calls for perf/accuracy to tl.dot
     qk_scale = 1.0
-    MATMUL_PRECISION = tl.float16
+    MATMUL_PRECISION = tl.float32
```
Comment (drisspg) [inline torch/_inductor/kernel/templated_attention.py:53]: this upcasts the bfloat16 /float16 case ?

File: benchmarks/transformer/score_mod.py (hunk 1 above)
Comment (Chillee) [reply benchmarks/transformer/score_mod.py:201]: yeah, it makes the benchmarks take too long lol.

File: torch/_inductor/kernel/templated_attention.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/kernel/templated_attention.py:53]: yeah I was trying to get float32 to work, gimme a bit.

**Comment by Chillee** [comment]
If I enable the new options (ROWS_ASSUME_SAFE and SCORE_MOD_IS_LINEAR), perf on benchmark goes up to 90% of FAV2.

File: benchmarks/transformer/score_mod.py [hunk 4]
```diff
@@ -6,25 +6,29 @@
...
 def benchmark_torch_function_in_microseconds(func: Callable, *args, **kwargs) -> float:
     # warmup
     for _ in range(5):
         func(*args, **kwargs)
-    t0 = benchmark.Timer(
-        stmt="func(*args, **kwargs)",
-        globals={"args": args, "kwargs": kwargs, "func": func},
-    )
-    return t0.adaptive_autorange(min_run_time=0.1).median * 1e6
+    return do_bench(lambda: func(*args, **kwargs)) * 1e3
```
Comment (drisspg) [inline benchmarks/transformer/score_mod.py:26]: This hides kernel launch time overhead but one criticism I have a seen is the increased time to launch, hence why I preferred this.

File: torch/_inductor/kernel/templated_attention.py [hunk 5]
```diff
@@ -117,25 +121,26 @@ def sdpa_grid(batch_size, num_heads, num_queries, d_model, meta):
...
         ) | indent_except_first(2) }}
+        # TODO: In the case that score_mod is linear, this can be LICMed
+        if not SCORE_MOD_IS_LINEAR:
+            qk *= 1.44269504
         # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
         # -- compute scaling constant ---
         row_max = tl.max(qk, 1)
         m_i_new = tl.maximum(m_i, row_max)
         masked_out_rows = (m_i_new == float("-inf"))
```
Comment (drisspg) [inline torch/_inductor/kernel/templated_attention.py:132]: Probs anna move this under the assume safe

File: benchmarks/transformer/score_mod.py [hunk 6]
```diff
@@ -6,25 +6,29 @@
...
+    # t0 = benchmark.Timer(
```
Comment (drisspg) [inline benchmarks/transformer/score_mod.py:27]: maybe remove this

**Review by drisspg — APPROVED** [review]
Some small nits, perf go ↑

**Comment by fmassa** [comment]
cc @danthe3rd , as IIRC this optimization had potentially some implications wrt accuracy

**Comment by Chillee** [comment]
@fmassa I guess we'll probably have a variety of configs for controlling it haha.

**Pull request closed without merging**